### PR TITLE
Feature/d3 asim 3398 Deprecated direct offer bid handling between client and d3a

### PR DESCRIPTION
**Pull request by hannesdiedrich**

### CODE REVIEW

File: d3a_api_client/setups/test_create_aggregator.py
Comment (fievelk) [inline d3a_api_client/setups/test_create_aggregator.py:33]:
```suggestion
                "current_market_fee: "
```

File: d3a_api_client/setups/test_create_aggregator.py
Comment (fievelk) [inline d3a_api_client/setups/test_create_aggregator.py:35]: Minor syntactic sugar:
```suggestion
            if not area_dict.get("asset_info"):
```

Comment (fievelk) [reply d3a_api_client/setups/test_create_aggregator.py:35]: Ah sorry, maybe you wanted to have a different behaviour when `asset_info` is `{}`?

File: d3a_api_client/setups/test_create_aggregator.py
Comment (fievelk) [inline d3a_api_client/setups/test_create_aggregator.py:40]: Just wanted to mention that `0` is considered to have a `False` value. So `key_in_dict_and_not_none_and_greater_than_zero` would be the same as evaluating `.get()`:
```python
area_dict = {'asset_info': {'available_energy_kWh': 0}}
if area_dict['asset_info'].get('available_energy_kWh'):
     print('true!')
else:
    print('false!')

false!
```
and
```python
area_dict = {'asset_info': {'available_energy_kWh': 1}}
if area_dict['asset_info'].get('available_energy_kWh'):
     print('true!')
else:
    print('false!')

true!
```

File: d3a_api_client/setups/test_create_aggregator.py
Comment (fievelk) [inline d3a_api_client/setups/test_create_aggregator.py:47]: This is unrelated to the current PR: I've seen this repeated in several scripts. Maybe it would help if we created a separate method that makes the intention more explicit. Something like: 
```suggestion
                if self.has_available_energy(area_dict):
```
and the checks happen inside the method.

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [inline d3a_api_client/redis_client_base.py:53]: Since we're deleting these lines, I think the `_generate_command_response_callback()` method can also be removed, right?
https://github.com/gridsingularity/d3a-api-client/blob/7fc1c0712c934022b16a6912891dab744e777b0b/d3a_api_client/redis_client_base.py#L107-L123

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [inline d3a_api_client/redis_client_base.py:28]: I would suggest to rename this to `RedisClientBase` to have the same name of the module (makes it easier to search in the project).

File: d3a_api_client/rest_market.py
Comment (fievelk) [inline d3a_api_client/rest_market.py:20]: Good catch here!

File: d3a_api_client/rest_market.py
Comment (fievelk) [inline d3a_api_client/rest_market.py:38]: Do you think it would make sense to also change the parameter name on the d3a-web side? From `device_uuid` to `area_uuid`?

File: integration_tests/steps/device.py
Comment (fievelk) [inline integration_tests/steps/device.py:28]:
```suggestion
    system("docker run -d --name d3a-tests --env REDIS_URL=redis://redis.container:6379/ "
```

File: d3a_api_client/setups/test_create_aggregator.py
Comment (hannesdiedrich) [reply d3a_api_client/setups/test_create_aggregator.py:47]: Yes, I reused your `_can_place_bid` and `_can_place_offer` in the integrationtests, but not here.
I agree, that it makes the intention more clear. 
We should review all `test_*` scripts in `/setups` and think about how user-friendly they are. For now i thought these are only for us for testing, but really they should be "good" examples. 
Do you think this is worth a separate task?

File: integration_tests/device.feature
Comment (fievelk) [inline integration_tests/device.feature:5]: Why are these changes necessary? Were there issues with the synchronization with d3a?

File: d3a_api_client/setups/test_create_aggregator.py
Comment (fievelk) [reply d3a_api_client/setups/test_create_aggregator.py:47]: Actually, I think this is what happens in `test_aggregator_ess.py` with `_can_place_offer` and `_can_place_bid`.

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:53]: You are totally right. Also in the redis market (we dont need the subscription to the responses of `"market_stats", "grid_fees", "dso_market_stats"` any more. Its handled in the batch commands. Thanks!

File: d3a_api_client/setups/test_create_aggregator.py
Comment (fievelk) [reply d3a_api_client/setups/test_create_aggregator.py:47]: > We should review all test_* scripts in /setups and think about how user-friendly they are. For now i thought these are only for us for testing, but really they should be "good" examples.
Do you think this is worth a separate task?

Yes, I think you're definitely right. But I wouldn't do this in the current PR: having a separate task as you suggest would be better in my opinion.

File: d3a_api_client/setups/test_create_aggregator.py
Comment (hannesdiedrich) [reply d3a_api_client/setups/test_create_aggregator.py:35]: No, you are right, changing it.

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:28]: I agree! Will do it!

File: integration_tests/device.feature
Comment (hannesdiedrich) [reply integration_tests/device.feature:5]: Yes, at least this is what i suspect. I tried to eliminate the non-deterministic behaviour of the tests as well as making them shorter. They were only working before, because they were running for a long time. Eventually, they received a market_cycle event. But if you run the d3a only with `-d 4h` and without slowdown, the market_cycle event will receive the d3a-api-client only when the d3a is already finished. 
I hope this makes sense. Happy to explain more.

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:28]: Actually, I dont see the need that these should be two classes. The potential `RedisClientBase` will not be used anywhere else. (only in auto_offer_bid_on_every_market.py and d3a_api_client/setups/auto_offer_bid_on_device.py which I will deleted also because there are not needed any more) 
EDIT: Actually, the future `RedisClientBase` could be reused in `RedisMarketClient`, too. OI wonder how this wasn't done before. I will probably see when implementing.

File: d3a_api_client/rest_market.py
Comment (hannesdiedrich) [reply d3a_api_client/rest_market.py:38]: Probably, yes. `area_uuid` is more general and as do not expose this to the user, it would be better. Will change it in the d3a-web.

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [reply d3a_api_client/redis_client_base.py:28]: When you say "two classes" you mean `RedisClient` and `RedisDeviceClient`, right? If that's the case, I agree that refactoring them into one class seems to be good give our current use.

Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:28]: I am currently trying to have one Baseclass `RedisClientBase` that works as Superclass for both `RedisDeviceClient` and `RedisMarketClient`. This would be the most clear way IMO.

Comment (fievelk) [reply d3a_api_client/redis_client_base.py:28]: Ah ok! Now I get it 👍

File: d3a_api_client/__init__.py
Comment (fievelk) [inline d3a_api_client/__init__.py:23]: Why did you opt for removing the parameter here? Wouldn't it be better to keep the same signature for the method across interface and implementations?

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [inline d3a_api_client/redis_client_base.py]: I have the feeling this method might be removed since it's just returning `self.area_slug`. The f-string here shouldn't be needed in any case, if slugify is returning a string. What do you think?

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [inline d3a_api_client/redis_client_base.py:52]: I'm not sure whether we could use it, but I just found this on the redis gh page:
> If all string responses from a client should be decoded, the user can specify `decode_responses=True` to `Redis.__init__`. In this case, any Redis command that returns a string type will be decoded with the encoding specified.

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [inline d3a_api_client/redis_client_base.py:53]: What kind of object is `self.pubsub.patterns[b'aggregator_response']`? Is it a callable?

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [inline d3a_api_client/redis_client_base.py:62]: Nit: this would probably deserve a separate method, to explain better what we're doing in this loop.

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [inline d3a_api_client/redis_client_base.py:69]: Transactions are never popped out of `self._transaction_id_buffer` in the new `RedisClientBase`, right?

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [inline d3a_api_client/redis_client_base.py:137]: Nit: it would be better to explicitly raise a custom exception rather than using this strategy to raise an `AssertionError`. Since this is mostly refactoring of existing code, it's not necessary to deal with this in this PR anyway.

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [inline d3a_api_client/redis_client_base.py:145]: Nit: unnecessary f-string (and single quotes)

File: d3a_api_client/redis_device.py
Comment (fievelk) [inline d3a_api_client/redis_device.py:19]: I think that `on_event_or_response` is not defined as an `@abstractmethod` on the interface, so it might be possible to remove it from here and from `RedisClientBase`. Or we might declar it as abstract on the interface, so that it will be clear that we want that method to be defined everywhere.

Comment (fievelk) [reply d3a_api_client/redis_device.py:19]: Actually, it looks like this method doesn't have any implementation anywhere. Couldn't we just remove it?

**Review by fievelk — APPROVED** [review]
LGTM in general. Just a few questions and minor points.
Two broader notes if you do some further refactoring: there are some f-strings that could be normal strings, and some single-quotes that might be converted to double-quotes.

File: d3a_api_client/__init__.py
Comment (hannesdiedrich) [reply d3a_api_client/__init__.py:23]: You are right, remnant of refactorization. Changing it back

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py]: You are right, will remove it and will use the area_slug in the channel_subs function directly.

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:53]: Sorry, this was messed up by me in the refactorization process. I will fix the whole `_subscribe_to_response_channels` actually.

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:62]: I wanted to do it, but then I searched for a name of  the function. It would be the same as the function here `_check_buffer_message_matching_command_and_id` . So this is the main part of the functionality. So I vote for leaving it as is.

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:69]: Indeed! I fix this in the frame of my comment above https://github.com/gridsingularity/d3a-api-client/pull/134#discussion_r618159818

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:137]: Indeed, will add it.

File: d3a_api_client/redis_device.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_device.py:19]: Indeed, you are right with your last comment. This in case the user likes to react in every event or repsonse.

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:52]: Hmm, I dont mind this being here. Also I found this comment: https://github.com/andymccurdy/redis-py/issues/1006#issuecomment-820626776

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [reply d3a_api_client/redis_client_base.py:53]: Since tests didn't fail because of this, it means that these functions are currently not tested, right? It would be great to do it so that we can spot errors automatically in the future.

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [reply d3a_api_client/redis_client_base.py:62]: Yeah, this is actually tricky. Maybe we could improve the readability of the method by turning the negative statement into a positive one. So, rather than:
```python
if not any(...):
    raise RedisAPIException(...)
```
we could have:
```python
if all(...):
    return True
else:
    raise RedisAPIException(...)
```
This is sometimes referred as the "Avoid negative conditionals" rule, and I think it helps readability.
Anyway, this is minor so feel free to leave it as it is if you want.

Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:62]: @fievelk `not any() != all()` 
We only want to check whether the condition is true for at least one of the responses. Not for all. So i dont see room for improvement here.

File: d3a_api_client/redis_client_base.py
Comment (hannesdiedrich) [reply d3a_api_client/redis_client_base.py:53]: Good catch! This error was introduced because of a wrong implementation of `_check_transaction_id_cached_out` which should be
```
    def _check_transaction_id_cached_out(self, transaction_id):
        return transaction_id not in self._transaction_id_buffer
```
Nobody complained, because the registration process was alway happening. Even though the response was not handled correctly.
As we already test all this code in the integrationtests, I can not think of a way to test this specific case. 
I fixed the above mentioned function and also some logging in the redis client. Should be working now. 
let me know if you have an Idea how to test it. Thanks a lot!

File: d3a_api_client/redis_client_base.py
Comment (fievelk) [reply d3a_api_client/redis_client_base.py:64]: I think my example should work though. It's not just replacing `not any()` with `all()`, it's also moving the exception into the `else` statement.
So we say:
> "if all the messages match the condition, then return True. Otherwise, raise an exception"

The version with `not any()` is saying:
> "if there is not even one message that matches the condition, then we raise an exception."

I hope I'm not getting confused by all the negations :P

Comment (fievelk) [reply d3a_api_client/redis_client_base.py:64]: Sorry, I was getting confused earlier. In this case, I would just use
```python
if any(...):
    return True
else:
    raise RedisAPIException(...)
```
to avoid negative conditionals. Again, it doesn't need to be changed here but it might be useful in the future. I find these conditionals clearer than the negative ones.

File: integration_tests/test_aggregator_ess.py
Comment (Hassan754) [inline integration_tests/test_aggregator_ess.py:17]: A reusable class for these tests defining all the repetitive properties would be great

Comment (hannesdiedrich) [reply integration_tests/test_aggregator_ess.py:17]: Great idea! Will do that.

**Review by Hassan754 — APPROVED** [review]
LGTM, Thanks a lot

**Review by spyrostz — DISMISSED** [review]
LGTM, also make sure to adapt the README once this is merged!

**Comment by hannesdiedrich** [comment]
> LGTM, also make sure to adapt the README once this is merged!

Let me do it now actually. Thanks!

**Review by spyrostz — APPROVED** [review]
LGTM

**Review by Hassan754 — APPROVED** [review]

**Pull request merged**
